### PR TITLE
Enable consumption of user-defined conversion operators declared in interfaces

### PR DESCRIPTION
**Pull request by AlekseyTs**
Related to https://github.com/dotnet/csharplang/pull/4773.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
@333fred, @RikkiGibson, @dotnet/roslyn-compiler Please review.

File: src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs [hunk 1]
```diff
@@ -23984,5 +23984,1108 @@ class C2 : I1<C2>
...
+}
+";
+            var compilation1 = CreateCompilation(source1, options: TestOptions.DebugDll,
+                                                 parseOptions: TestOptions.RegularPreview,
+                                                 targetFramework: TargetFramework.NetCoreApp);
+
+            var error = (op == "explicit" ? ErrorCode.ERR_NoExplicitConv : ErrorCode.ERR_NoImplicitConv);
+
+            compilation1.VerifyDiagnostics(
+                // (8,16): error CS0030: Cannot convert type 'I1<T>' to 'int'
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:24031]: We might want to consider a better diagnostic experience here. Perhaps let the conversion exist, but make it invalid?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Symbol/Symbols/StaticAbstractMembersInInterfacesTests.cs:24031]: > We might want to consider a better diagnostic experience here. Perhaps let the conversion exist, but make it invalid?

The proposed specification doesn't consider any user-defined conversion operators from interface types. If specification changes, things could change.

**Review by 333fred — APPROVED** [review]
LGTM (commit 1)

**Comment by AlekseyTs** [comment]
@RikkiGibson, @dotnet/roslyn-compiler For the second review.

**Comment by AlekseyTs** [comment]
@RikkiGibson, @dotnet/roslyn-compiler For the second review.

**Comment by RikkiGibson** [comment]
Taking a look now.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs [hunk 2]
```diff
@@ -382,6 +382,25 @@ internal ImmutableArray<Conversion> UnderlyingConversions
             }
         }
 
+        internal TypeParameterSymbol? ConstrainedToTypeOpt
+        {
+            get
+            {
+                var uncommonData = _uncommonData;
+                if (uncommonData != null && uncommonData._conversionMethod is null)
```
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs:390]: > I didn't understand why we need the _conversionMethod to be null here to enter the block.

I am following the pattern esteblished in the previous property. If _conversionMethod is not null, we never look at the _conversionResult. Presumably both are never set at the same time, but I don't want to take any chances.

**Pull request merged**